File: code/13.罗马数字转整数.go
/*
 * @lc app=leetcode.cn id=13 lang=golang
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
func romanToInt(s string) int {
	m := make(map[byte]int)
	m['I'] = 1
	m['V'] = 5
	m['X'] = 10
	m['L'] = 50
	m['C'] = 100
	m['D'] = 500
	m['M'] = 1000

	tmp := 0
	res := 0
	for i := 0; i < len(s); i++ {
		if i+1 < len(s) && m[s[i]] < m[s[i+1]] {
			tmp = m[s[i+1]] - m[s[i]]
			i++
		} else {
			tmp = m[s[i]]
		}
		res += tmp
	}
	return res
}

// @lc code=end

File: code/13.罗马数字转整数.cpp
/*
 * @lc app=leetcode.cn id=13 lang=cpp
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
class Solution {
public:
    int romanToInt(string s) {
        if(s.size() == 0) {
            return 0;
        }

        unordered_map<char, int> map;
        map['I'] = 1;
        map['V'] = 5;
        map['X'] = 10;
        map['L'] = 50;
        map['C'] = 100;
        map['D'] = 500;
        map['M'] = 1000;

        int res = 0;
        int tmp = 0;

        for(int i = 0; i < s.size(); ++i) {
             if(i+1 < s.size() && map[s[i]] < map[s[i+1]]) {
                 tmp = map[s[i+1]] - map[s[i]];
                 ++i;
             } else {
                 tmp = map[s[i]];
             }
             res += tmp;
        }

        return res;
    }
};
// @lc code=end

